crates/bevy_ecs/src/ptr.rs

```diff
@@ -0,0 +1,62 @@
+use std::cell::UnsafeCell;
+
+/// A "reference holder" similar to [`UnsafeCell`] except:
+///
+/// - it can be constructed from a shared or mutable reference
+/// - references to it can be shared between threads
+///
+/// `bevy_ecs` uses this to have a single argument type in several core `unsafe` [`World`](crate::world::World)
+/// access methods. In those contexts, the multi-threaded executor ensures that the active references
+/// of systems do not alias in unsound ways.
+pub enum SemiSafeCell<'a, T> {
+    Ref(&'a T),
+    Mut(&'a UnsafeCell<T>),
+}
```

Comment:
(I can't see the text box to reply to reviews and support suggested leaving a comment might fix it).